### PR TITLE
Generate bitflags and enums in high-level crates

### PR DESCRIPTION
**Pull request by gkoz**
Before generating any more crates I'd like to start generating enums.

Some things to decide there are:

* Do we want to list all generated enums in the config? We probably should either change "object" to something more generic or add `enumeration`, `record` etc.

* The `__Nonexhaustive` variant protects from breakage caused by adding variants. This implementation always adds it but we may want a different policy. Adding `__Nonexhaustive` everywhere could potentially be a serious breaking change for gtk.

* The working assumption here is that in the future the sys crates won't have any enums or bitflags, just `c_int` and `c_uint` newtypes or type aliases and constants. But perhaps not?

Relevant issues: #108, #124.

### CODE REVIEW

**Comment by EPashkin** [comment]
IMHO we can safe use `object` in config as `object to process` not as `decendant of GObject` etc (as we do with interfaces and records), so separating unneeded.
Only if "subtype" need some unique parameters may be reason to separating.

File: src/codegen/enums.rs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+use ffi;
+use glib_ffi;
+use glib::error::ErrorDomain;
+use glib::translate::*;
+"#));
+
+        for (_, type_) in env.library.namespace_types(namespaces::MAIN) {
+            if let Type::Enumeration(ref enum_) = *type_ {
+                try!(generate_enum(env, w, enum_));
+                if let Some (cfg) = version_condition_string(env.config.min_cfg_version,
```
Comment (EPashkin) [inline src/codegen/enums.rs]: Too many calls `version_condition_string` with these parameters, maybe its time to add new function with parameters `(env: &Env, version: Option<Version>)`?

File: src/codegen/enums.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+use std::path::Path;
+use traits::*;
+use version::Version;
+
+pub fn generate(env: &Env, root_path: &Path, mod_rs: &mut Vec<String>) {
+    let path = root_path.join("enums.rs");
+    mod_rs.push("\nmod enums;".into());
+    file_saver::save_to_file(path, env.config.make_backup, |w| {
+        try!(general::start_comments(w, &env.config));
+        try!(writeln!(w, "{}", r#"
```
Comment (EPashkin) [inline src/codegen/enums.rs]: This really needed be raw string? For enums.rs changing all raw to normal seems has no effect on result file.
Also mixing multiline strings starting from beginning of line and not slightly confusing.

**Comment by EPashkin** [comment]
Don't see problem in `__Nonexhaustive`.

**Comment by gkoz** [comment]
I'll clean those things up. And will add status checks against config.

**Comment by gkoz** [comment]
Updated. Not sure if it's possible to make multiline literals prettier.

Bitfields and enums are ignored by default now and need to be listed in the config.

File: src/codegen/enums.rs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                    c.type_id.map_or(false, |tid| tid.ns_id == namespaces::MAIN)
+            });
+        let mut first = true;
+        for config in configs {
+            if let Type::Enumeration(ref enum_) = *env.library.type_(config.type_id.unwrap()) {
+                if first {
+                    mod_rs.push("\nmod enums;".into());
+                    first = false;
+                }
+                if let Some (cfg) = version_condition_string(env, enum_.version, false, 0) {
```
Comment (EPashkin) [inline src/codegen/enums.rs]: I thinking about `version_condition_string_default(env, enum_.version)` here, as too many `(... ,false, 0)` in project.

File: src/codegen/general.rs [hunk 4]
```diff
@@ -86,19 +85,17 @@ pub fn define_boxed_type(w: &mut Write, type_name: &str, glib_name: &str,
...
+                         indent: usize) -> Result<()> {
+    if let Some(s) = version_condition_string(env, version, commented, indent) {
         try!(writeln!(w, "{}", s));
     }
     Ok(())
 }
 
-pub fn version_condition_string(min_cfg_version: Version,
-        version: Option<Version>, commented: bool, indent: usize) -> Option<String> {
+pub fn version_condition_string(env: &Env, version: Option<Version>, commented: bool, indent: usize)        -> Option<String> {
```
Comment (EPashkin) [inline src/codegen/general.rs]: Too long

File: src/codegen/objects.rs [hunk 5]
```diff
@@ -18,7 +18,7 @@ pub fn generate(env: &Env, root_path: &Path, mod_rs: &mut Vec<String>, traits: &
         let class_analysis = match info {
             Some(info) => info,
             None => {
-                warn!("Class or interface {} not found.", obj.name);
+                info!("Class or interface {} not found.", obj.name);
```
Comment (EPashkin) [inline src/codegen/objects.rs]: IMHO if config states that type must generated and no such type, then it sure warning and maybe ever error,
and better seen in run log and not when checking file list later.

File: src/codegen/enums.rs [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                enum_.name, member.name));
+        }
+        if has_failed_member {
+            try!(writeln!(w, "\t\t\t_ => Some({}::Failed),", enum_.name));
+        } else {
+            try!(writeln!(w, "\t\t\t_ => Some({}::__Nonexhaustive),", enum_.name));
+        }
+
+        try!(writeln!(w, "{}",
+"        }
```
Comment (EPashkin) [inline src/codegen/enums.rs]: `try!(writeln!(w, "\t\t}}\n\t}}\n}}\n\n"));` ?
Not sure if it really better.

File: src/codegen/enums.rs (hunk 3 above)
Comment (gkoz) [inline src/codegen/enums.rs]: Overall it seems the `tabs` parameter is used a lot and other ways of adding tabs are going to be a bit ugly too. Factoring out `commented` feels like diminishing returns or I'm just not seeing a clean way.

File: src/codegen/enums.rs (hunk 6 above)
Comment (gkoz) [inline src/codegen/enums.rs]: Too dense :)

**Comment by EPashkin** [comment]
@gkoz Merge when ready

**Comment by EPashkin** [comment]
When tried regenerate gtk after this `GesturePan` lose its `#[cfg(feature = "v3_14")]`

**Comment by EPashkin** [comment]
Ignore this, I just don't generated `Gtk.Orientation`

**Pull request merged**